File: LudoGame/LudoGame/Views/BoardView.swift
import SwiftUI

struct BoardView: View {
    let state: LudoGameState
    let selectableTokenIDs: Set<Int>
    let onTokenTap: (Int) -> Void

    var body: some View {
        GeometryReader { geo in
            let size = min(geo.size.width, geo.size.height)

            ZStack {
                boardBackground(size: size)
                pathCells(size: size)
                homeAreas(size: size)
                centerTriangle(size: size)
                tokensLayer(size: size)
            }
            .frame(width: size, height: size)
            .position(x: geo.size.width / 2, y: geo.size.height / 2)
        }
    }

    @ViewBuilder
    private func boardBackground(size: CGFloat) -> some View {
        RoundedRectangle(cornerRadius: size * 0.04)
            .fill(Color(red: 0.93, green: 0.94, blue: 0.96))
            .overlay(
                RoundedRectangle(cornerRadius: size * 0.04)
                    .stroke(Color.black.opacity(0.08), lineWidth: 2)
            )
    }

    @ViewBuilder
    private func homeAreas(size: CGFloat) -> some View {
        ForEach(PlayerColor.allCases) { color in
            homeArea(for: color, size: size)
        }
    }

    @ViewBuilder
    private func homeArea(for color: PlayerColor, size: CGFloat) -> some View {
        let rect = homeRect(for: color.homeCorner, size: size)
        RoundedRectangle(cornerRadius: size * 0.02)
            .fill(color.lightColor)
            .overlay(
                RoundedRectangle(cornerRadius: size * 0.02)
                    .stroke(color.color.opacity(0.8), lineWidth: 2)
            )
            .frame(width: rect.width, height: rect.height)
            .position(x: rect.midX, y: rect.midY)
    }

    @ViewBuilder
    private func pathCells(size: CGFloat) -> some View {
        ForEach(0..<LudoBoardLayout.pathLength, id: \.self) { index in
            let point = LudoBoardLayout.coordinate(atAbsoluteTrackIndex: index)
            let cellSize = size * 0.055
            let isSafe = PlayerColor.safePositions.contains(index)

            RoundedRectangle(cornerRadius: 3)
                .fill(isSafe ? Color.white : Color.white.opacity(0.85))
                .overlay(
                    RoundedRectangle(cornerRadius: 3)
                        .stroke(Color.black.opacity(0.08), lineWidth: 1)
                )
                .overlay {
                    if isSafe {
                        Image(systemName: "star.fill")
                            .font(.system(size: cellSize * 0.35))
                            .foregroundStyle(Color.orange.opacity(0.8))
                    }
                }
                .frame(width: cellSize, height: cellSize)
                .position(x: point.x * size, y: point.y * size)
        }
    }

    @ViewBuilder
    private func centerTriangle(size: CGFloat) -> some View {
        ZStack {
            ForEach(PlayerColor.allCases) { color in
                TriangleWedge()
                    .fill(color.color.opacity(0.85))
                    .frame(width: size * 0.22, height: size * 0.22)
                    .rotationEffect(rotation(for: color))
            }

            Circle()
                .fill(Color.white)
                .frame(width: size * 0.08, height: size * 0.08)
                .overlay(Circle().stroke(Color.black.opacity(0.1), lineWidth: 1))
        }
        .position(x: size * 0.5, y: size * 0.5)
    }

    @ViewBuilder
    private func tokensLayer(size: CGFloat) -> some View {
        ForEach(state.players) { player in
            ForEach(Array(player.tokens.enumerated()), id: \.element.id) { stackIndex, token in
                let point = LudoBoardLayout.coordinate(for: token, stackIndex: stackIndex)
                let tokenSize = size * 0.055
                let isSelectable = selectableTokenIDs.contains(token.id) && player.id == state.currentPlayer.id

                Button {
                    if isSelectable {
                        onTokenTap(token.id)
                    }
                } label: {
                    TokenView(
                        color: player.id,
                        isSelectable: isSelectable,
                        isCurrentPlayer: player.id == state.currentPlayer.id
                    )
                    .frame(width: tokenSize, height: tokenSize)
                }
                .buttonStyle(.plain)
                .disabled(!isSelectable)
                .position(x: point.x * size, y: point.y * size)
                .animation(.spring(response: 0.35, dampingFraction: 0.8), value: token.position)
            }
        }
    }

    private func homeRect(for corner: BoardCorner, size: CGFloat) -> CGRect {
        let side = size * 0.34
        switch corner {
        case .bottomLeft:
            return CGRect(x: size * 0.02, y: size - side - size * 0.02, width: side, height: side)
        case .topLeft:
            return CGRect(x: size * 0.02, y: size * 0.02, width: side, height: side)
        case .topRight:
            return CGRect(x: size - side - size * 0.02, y: size * 0.02, width: side, height: side)
        case .bottomRight:
            return CGRect(x: size - side - size * 0.02, y: size - side - size * 0.02, width: side, height: side)
        }
    }

    private func rotation(for color: PlayerColor) -> Angle {
        switch color {
        case .red: return .degrees(135)
        case .green: return .degrees(45)
        case .yellow: return .degrees(-45)
        case .blue: return .degrees(-135)
        }
    }
}

private struct TriangleWedge: Shape {
    func path(in rect: CGRect) -> Path {
        var path = Path()
        path.move(to: CGPoint(x: rect.midX, y: rect.midY))
        path.addLine(to: CGPoint(x: rect.maxX, y: rect.minY))
        path.addLine(to: CGPoint(x: rect.maxX, y: rect.maxY))
        path.closeSubpath()
        return path
    }
}

#Preview {
    BoardView(
        state: LudoGameState(playerCount: 4),
        selectableTokenIDs: [],
        onTokenTap: { _ in }
    )
    .padding()
}

File: LudoGameAndroid/app/src/main/java/com/ludogame/app/ui/components/BoardView.kt
package com.ludogame.app.ui.components

import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.drawscope.rotate
import androidx.compose.ui.unit.IntOffset
import androidx.compose.ui.unit.dp
import com.ludogame.app.engine.LudoBoardLayout
import com.ludogame.app.models.BoardCorner
import com.ludogame.app.models.LudoGameState
import com.ludogame.app.models.PlayerColor
import kotlin.math.roundToInt

@Composable
fun BoardView(
    state: LudoGameState,
    selectableTokenIds: Set<Int>,
    onTokenTap: (Int) -> Unit,
    modifier: Modifier = Modifier
) {
    BoxWithConstraints(modifier = modifier) {
        val boardSize = minOf(maxWidth, maxHeight)

        Box(
            modifier = Modifier
                .size(boardSize)
                .align(Alignment.Center)
        ) {
            Canvas(modifier = Modifier.fillMaxSize()) {
                val sizePx = size.minDimension
                val cornerRadius = sizePx * 0.04f

                drawRoundRect(
                    color = Color(0xFFEDEEF2),
                    cornerRadius = CornerRadius(cornerRadius, cornerRadius),
                    size = Size(sizePx, sizePx)
                )
                drawRoundRect(
                    color = Color.Black.copy(alpha = 0.08f),
                    cornerRadius = CornerRadius(cornerRadius, cornerRadius),
                    size = Size(sizePx, sizePx),
                    style = Stroke(width = 2f)
                )

                PlayerColor.allCases.forEach { color ->
                    val rect = homeRect(color.homeCorner, sizePx)
                    drawRoundRect(
                        color = color.lightColor,
                        topLeft = Offset(rect.left, rect.top),
                        size = Size(rect.width, rect.height),
                        cornerRadius = CornerRadius(sizePx * 0.02f, sizePx * 0.02f)
                    )
                    drawRoundRect(
                        color = color.composeColor.copy(alpha = 0.8f),
                        topLeft = Offset(rect.left, rect.top),
                        size = Size(rect.width, rect.height),
                        cornerRadius = CornerRadius(sizePx * 0.02f, sizePx * 0.02f),
                        style = Stroke(width = 2f)
                    )
                }

                repeat(LudoBoardLayout.PATH_LENGTH) { index ->
                    val point = LudoBoardLayout.coordinateAtAbsoluteTrackIndex(index)
                    val cellSize = sizePx * 0.055f
                    val isSafe = PlayerColor.safePositions.contains(index)
                    val topLeft = Offset(
                        x = point.x * sizePx - cellSize / 2f,
                        y = point.y * sizePx - cellSize / 2f
                    )

                    drawRoundRect(
                        color = if (isSafe) Color.White else Color.White.copy(alpha = 0.85f),
                        topLeft = topLeft,
                        size = Size(cellSize, cellSize),
                        cornerRadius = CornerRadius(3f, 3f)
                    )
                    drawRoundRect(
                        color = Color.Black.copy(alpha = 0.08f),
                        topLeft = topLeft,
                        size = Size(cellSize, cellSize),
                        cornerRadius = CornerRadius(3f, 3f),
                        style = Stroke(width = 1f)
                    )
                }

                val wedgeSize = sizePx * 0.22f
                PlayerColor.allCases.forEach { color ->
                    val rotation = when (color) {
                        PlayerColor.RED -> 135f
                        PlayerColor.GREEN -> 45f
                        PlayerColor.YELLOW -> -45f
                        PlayerColor.BLUE -> -135f
                    }
                    rotate(rotation, pivot = center) {
                        val path = Path().apply {
                            moveTo(center.x, center.y)
                            lineTo(center.x + wedgeSize / 2f, center.y - wedgeSize / 2f)
                            lineTo(center.x + wedgeSize / 2f, center.y + wedgeSize / 2f)
                            close()
                        }
                        drawPath(path, color.composeColor.copy(alpha = 0.85f))
                    }
                }

                drawCircle(
                    color = Color.White,
                    radius = sizePx * 0.04f,
                    center = center
                )
                drawCircle(
                    color = Color.Black.copy(alpha = 0.1f),
                    radius = sizePx * 0.04f,
                    center = center,
                    style = Stroke(width = 1f)
                )
            }

            state.players.forEach { player ->
                player.tokens.forEachIndexed { stackIndex, token ->
                    val point = LudoBoardLayout.coordinate(token, stackIndex)
                    val tokenSize = boardSize * 0.055f
                    val isSelectable = selectableTokenIds.contains(token.id) &&
                        player.id == state.currentPlayer.id

                    TokenView(
                        color = player.id,
                        isSelectable = isSelectable,
                        isCurrentPlayer = player.id == state.currentPlayer.id,
                        modifier = Modifier
                            .size(tokenSize)
                            .offset {
                                IntOffset(
                                    x = (point.x * boardSize.toPx() - tokenSize.toPx() / 2f).roundToInt(),
                                    y = (point.y * boardSize.toPx() - tokenSize.toPx() / 2f).roundToInt()
                                )
                            }
                            .then(
                                if (isSelectable) {
                                    Modifier.clickable { onTokenTap(token.id) }
                                } else {
                                    Modifier
                                }
                            )
                    )
                }
            }
        }
    }
}

@Composable
fun TokenView(
    color: PlayerColor,
    isSelectable: Boolean,
    isCurrentPlayer: Boolean,
    modifier: Modifier = Modifier
) {
    val infiniteTransition = rememberInfiniteTransition(label = "tokenPulse")
    val pulseScale by infiniteTransition.animateFloat(
        initialValue = 1f,
        targetValue = if (isSelectable) 1.25f else 1f,
        animationSpec = infiniteRepeatable(
            animation = tween(500),
            repeatMode = RepeatMode.Reverse
        ),
        label = "pulseScale"
    )

    Box(
        modifier = modifier,
        contentAlignment = Alignment.Center
    ) {
        Box(
            modifier = Modifier
                .fillMaxSize()
                .clip(CircleShape)
                .background(color.composeColor)
                .border(2.dp, Color.White, CircleShape)
        )

        if (isSelectable) {
            Box(
                modifier = Modifier
                    .fillMaxSize(pulseScale.coerceAtMost(1.3f))
                    .clip(CircleShape)
                    .border(3.dp, Color(0xFFFFD600), CircleShape)
            )
        }

        if (isCurrentPlayer) {
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .clip(CircleShape)
                    .border(1.dp, Color.Black.copy(alpha = 0.25f), CircleShape)
            )
        }
    }
}

private fun homeRect(corner: BoardCorner, size: Float): Rect {
    val side = size * 0.34f
    val inset = size * 0.02f
    return when (corner) {
        BoardCorner.BOTTOM_LEFT -> Rect(inset, size - side - inset, inset + side, size - inset)
        BoardCorner.TOP_LEFT -> Rect(inset, inset, inset + side, inset + side)
        BoardCorner.TOP_RIGHT -> Rect(size - side - inset, inset, size - inset, inset + side)
        BoardCorner.BOTTOM_RIGHT -> Rect(size - side - inset, size - side - inset, size - inset, size - inset)
    }
}
